model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java

```diff
@@ -37,11 +37,27 @@
         @NamedQuery(name="deleteExpiredUserSessions", query="delete from PersistentUserSessionEntity sess where sess.realmId = :realmId AND sess.offline = :offline AND sess.lastSessionRefresh < :lastSessionRefresh"),
         @NamedQuery(name="updateUserSessionLastSessionRefresh", query="update PersistentUserSessionEntity sess set lastSessionRefresh = :lastSessionRefresh where sess.realmId = :realmId" +
                 " AND sess.offline = :offline AND sess.userSessionId IN (:userSessionIds)"),
-        @NamedQuery(name="findUserSessionsCount", query="select count(sess) from PersistentUserSessionEntity sess where sess.offline = :offline"),
+        @NamedQuery(name="findUserSessionsCount", query="select count(sess) from PersistentUserSessionEntity sess where sess.offline = :offline and (sess.realmId = :realmId or :realmId is null)"),
```

Comment:
Is there any valid way to have `realmId == NULL`? Any session belongs to a realm, so if `realmId` is `null`, this would indicate a bug.
```suggestion
        @NamedQuery(name="findUserSessionsCount", query="select count(sess) from PersistentUserSessionEntity sess where sess.offline = :offline and sess.realmId = :realmId"),
```

This seems to stem from the change in `getUserSessionsCount(boolean offline)` that now delegates to `getUserSessionsCount(realm, offline)` and passes in `null` in `realm`. This change should be reverted, as the condition `realmID IS null` should lead to empty result.

The change is only needed in the case of the newly introduced `int getUserSessionsCount(boolean offline, ClientModel clientModel);` variant that needs to have `RealmModel` as well.

Reply:
realm==null means that the count of all user sessions is returned. This is currently used by [OfflinePersistentUserSessionLoader](https://github.com/keycloak/keycloak/blob/8432513daa1c19437b48b2a87dcc97e093212008/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/initializer/OfflinePersistentUserSessionLoader.java#L67).

Reply:
That's correct, but the query this way can effectively cause a database to perform evaluation of the `OR` condition instead of using index lookup. For that reason, two separate queries for this two cases would be preferrable.

Reply:
Okay, I'll check that and add a dedicated query if necessary.

Reply:
I did some tests with `and sess.realmId = coalesce(:realmId, sess.realmId)` it seems to always use the proper index.